File: android/app/src/main/java/com/bluelink/android/session/QueuedRouteGate.kt
package com.bluelink.android.session

import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CancellationException

internal class QueuedRouteGate {
    private var state=0 // waiting, switching, started, stopped
    private var decision: CompletableDeferred<Boolean>?=null
    @Synchronized fun request(): CompletableDeferred<Boolean>? {
        if(state != 0) return null
        state=1
        return CompletableDeferred<Boolean>().also { decision=it }
    }
    @Synchronized fun resolve(accepted: Boolean) {
        if(state != 1) return
        state=if(accepted) 3 else 0
        decision!!.complete(accepted)
    }
    suspend fun start() {
        while(true) {
            val pending=synchronized(this) {
                when(state) {
                    0 -> { state=2; return }
                    1 -> requireNotNull(decision)
                    else -> throw CancellationException("USB queue ended")
                }
            }
            pending.await()
        }
    }
}

File: windows/BlueLink.App/Session/QueuedRouteGate.cs
namespace BlueLink.Session;

// Serializes the last possible route change with opening the transfer stream.
internal sealed class QueuedRouteGate
{
    private readonly object _gate = new();
    private int _state; // waiting, switching, started, stopped
    private TaskCompletionSource<bool>? _decision;
    public Task<bool>? Request()
    {
        lock (_gate)
        {
            if (_state != 0) return null;
            _state = 1;
            return (_decision = new(TaskCreationOptions.RunContinuationsAsynchronously)).Task;
        }
    }
    public void Resolve(bool accepted)
    {
        lock (_gate)
        {
            if (_state != 1) return;
            _state = accepted ? 3 : 0;
            _decision!.TrySetResult(accepted);
        }
    }
    public async Task StartAsync(CancellationToken token)
    {
        while (true)
        {
            Task pending;
            lock (_gate)
            {
                token.ThrowIfCancellationRequested();
                if (_state == 0) { _state = 2; return; }
                if (_state != 1) throw new OperationCanceledException("USB 队列已结束", token);
                pending = _decision!.Task;
            }
            await pending.WaitAsync(token);
        }
    }
}
